src/unity/python/turicreate/toolkits/activity_classifier/util.py

```diff
@@ -68,7 +70,43 @@ def random_split_by_session(dataset, session_id, fraction=0.9, seed=None):
             'Input "dataset" must contain a column called %s.' % session_id)
 
     unique_sessions = _SFrame({'session': dataset[session_id].unique()})
+    if len(unique_sessions) < _MIN_NUM_SESSIONS_FOR_SPLIT:
+        print ("The dataset has less than the minimum of", _MIN_NUM_SESSIONS_FOR_SPLIT, "sessions required for train-validation split. Continuing without validation set")
+        return dataset, None
+
     chosen, not_chosen = unique_sessions.random_split(fraction, seed)
-    train = dataset.filter_by(chosen['session'], session_id)
-    valid = dataset.filter_by(not_chosen['session'], session_id)
-    return train, valid
+
+    def split_using_is_in(data, session_id, not_chosen):
+        not_chosen_filter = data[session_id].is_in(not_chosen['session'])
+
+        train = data[1 - not_chosen_filter]
+        valid = data[not_chosen_filter]
```

Comment:
All of this logic makes sense, but it's frustrating, since random_split works by generating a random Boolean value for each row and then performing a filter internally, before returning the two resulting SArrays. We're having to re-implement so much of this operation, and the is_in case requires a linear search through `not_chosen` for each row, and the filter case requires sorting.

Could we somehow generate a random boolean for each row that yields the same result for each session ID? In principle, all it would take is a suitable hash function:

```
chosen_filter = data[session_id].apply(lambda x: hash_fn(x + seed) < fraction * MAX_HASH_FN_OUTPUT)
train = data[chosen_filter]
valid = data[1 - chosen_filter]
```

The total operation should then be linear, at face value. All that is required is a suitable hash function. For `hash_fn(x)` it may even suffice to re-initialize a RNG with `x` as the seed, perform one draw from uniform(0,1) and compare against `fraction`.

If there's something I'm missing, or if Python doesn't really make it easy enough to do the above, I could always investigate this approach as a followup someday. How long does this approach take in practice?

Reply:
Hmm - interesting. This sounds like a good idea.  
My only concern is that it only guarantees that the sessions are actually divided according to the divided fraction in expectancy. For a small number of sessions (which is the typical case) - the actual fraction might be quite different.  
I'm not sure how accurate our current `random_split` mechanism is (which we use to divide the sessions). Maybe @ylow has some more information.

I have a bolt task where I've compared the `is_in` and the `filter` implementations for various dataset sizes (number of sessions). I'll run it again using your idea - and compare run time and fraction accuracy.

Reply:
Ah, good point about the variance in the split. Looking at the implementation: [unity_sframe::random_split](https://github.com/apple/turicreate/blob/e43e2133d725cf8cbdb3d92f8d4d46facb58a3aa/src/unity/lib/unity_sframe.cpp#L1071) calls into [unity_sarray::make_uniform_boolean_array](https://github.com/apple/turicreate/blob/e43e2133d725cf8cbdb3d92f8d4d46facb58a3aa/src/unity/lib/unity_sarray.cpp#L1690), which uses essentially the idea above.

Interestingly, the C++ version has an `exact` parameter that causes it split as close as possible to the desired proportion. But that implementation also uses sorting. I would think there's a more efficient way to choose a random k-subset of n elements. I wonder if we could just use [numpy.random.choice](https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.choice.html). Something like:
```
validation_session_ids = numpy.random.choice(num_sessions, fraction*num_sessions, replace=False)
# Use some numpy magic I don't know to convert to an array of length num_sessions with 1 in each row indicated by validation_session_ids
# Convert to filter SArray
```

Reply:
Good idea - but it goes beyond the scope of this PR.  
Let's open a separate issue for that?

Reply:
BTW - session_id may be a string - so using `random.seed(x) ; random.uniform(0,1)` as the "hash" function won't work. At least not out of the box without pre-processing (in the string case).  
I'll try to see if I can find some more suitable hash func.